File: 卡码/代码随想录/1/02-链表/C++/04-两两交换链表节点/LeetCode_0024.cpp
//24. 两两交换链表中的节点
struct ListNode{
	int val;
	ListNode* next;
	ListNode():val(-1),next(nullptr){}
	ListNode(int val):val(val),next(nullptr){}
	ListNode(int val,ListNode* next):val(val),next(next){}
};

class Solution {
public:
  ListNode* swapPairs(ListNode* head) {
  	//如果链表为空或者只有一个元素
		if(head==nullptr||head->next==nullptr) return head;
		ListNode* dummy=new ListNode(-1,head);
		ListNode* current=dummy;
		while(current->next&&current->next->next){
			ListNode* tmp=current->next;	//保留一号位
			ListNode* tmp1=tmp->next->next;	//保留三号位
			current->next=tmp->next;
			current->next->next=tmp;
			tmp->next=tmp1;
			current=tmp;
		}
		return dummy->next;
  }
};

File: 卡码/代码随想录/1/02-链表/Go/04-两两交换链表节点/LeetCode_0024.go
//24. 两两交换链表中的节点
package main

type ListNode struct{
	Val int
	Next *ListNode
}

func swapPairs(head *ListNode) *ListNode {
	//链表为空或者只有一个元素
	if head==nil||head.Next==nil {
		return head
	}
	dummy:=&ListNode{-1,head}
	current:=dummy
	for current.Next!=nil&&current.Next.Next!=nil {
		tmp:=current.Next	//保留1号位
		tmp1:=tmp.Next.Next	//保留3号位
		current.Next=tmp.Next
		current.Next.Next=tmp
		tmp.Next=tmp1
		current=tmp
	}
	return dummy.Next
}
